lib/ttt.ex

```diff
@@ -5,55 +5,133 @@ defmodule TTT do
+      choice == "1" -> player_player()
+      choice == "2" -> player_comp()
+      choice == "3" -> comp_comp()
+    end
+  end
+
+  def start({player1, player2} = _players) do
     board = Board.new
-    game = Game.new
+    game = Game.new(player1, player2)
     status = Game.status(game)
     Board.show(board)
-    play(board, game, status, :x)
+    play(board, game, status, player1)
+  end
+
+  def player_player do
+    p1name = set_player_name(1)
+    p2name = set_player_name(2)
+    player1 = Player.new(p1name, :x, :human)
+    player2 = Player.new(p2name, :o, :human)
+    {player1, player2}
+  end
+
+  def player_comp do
+    p1name = set_player_name(1)
+    p2name = set_player_name(2)
+    player1 = Player.new(p1name, :x, :human)
+    player2 = Player.new(p2name, :o, :computer)
+    {player1, player2}
+  end
+
+  def comp_comp do
+    p1name = set_player_name(1)
+    p2name = set_player_name(2)
+    player1 = Player.new(p1name, :x, :computer)
+    player2 = Player.new(p2name, :o, :computer)
+    {player1, player2}
   end
 
-  def play(board, game, status, turn) when status == :underway do
-    input = get_input(turn)
-    move = match_input(String.trim(input))
-    with {:ok, game} <- Game.play_turn(game, turn, move) do
+  def play(board, game, status, %Player{name: _name, token: _token, type: type} = turn)
+    when status == :underway do
+    cond do
+      type == :human ->
+        input = get_move(turn)
+        move = match_input(String.trim(input))
+        make_a_move(board, game, turn, move)
+      type == :computer ->
+        move = generate_move(game, 1)
+        make_a_move(board, game, turn, move)
+    end
+  end
+
+  def play(_board, _game, {_progress, {outcome, person}} = status, _turn) when status != :underway do
+    IO.puts "Game over!\nThe #{outcome} is #{person}!"
+  end
+
+  def play(_board, _game, {_progress, outcome} = status, _turn) when status != :underway do
+    IO.puts "Game over!\nIt's a #{outcome}!"
+  end
+
+  def make_a_move(board, game, %Player{name: _name, token: token, type: _type} = turn, move) do
+    with {:ok, game} <- Game.play_turn(game, token, move) do
       update_visual(board, game)
       status = Game.status(game)
-      turn = switch_turn(turn)
+      turn = switch_turn(game, token)
       play(board, game, status, turn)
     else
       {:error, error} -> IO.puts "\nInvalid move: #{error}. Please try again. \n"
-      play(board, game, status, turn)
+      input = get_move(turn)
+      move = match_input(String.trim(input))
+      make_a_move(board, game, turn, move)
     end
   end
 
-  def play(_board, _game, {_progress, {outcome, person}} = status, _turn) when status != :underway do
-    IO.puts "Game over!\n#{outcome} is #{person}!"
+  def generate_move(%Game{turns: %{x: p1moves, o: p2moves}} = game, input) do
+    input_string = Integer.to_string(input)
+    move = match_input(input_string)
+    cond do
+      MapSet.member?(p1moves, move) or MapSet.member?(p2moves, move) ->
```

Comment:
I am not sure how the logic here works, exactly... Is there a way to simplify this? Or extract it into a few well-named methods?

Reply:
Since I already have a function called match_input that will match inputs such as 1,2,3... to coordinates {0,0},{1,0}... etc. I simply use a function called generate_move to start simulating human inputs started at 1, which translate into {0,0}, then the conditional statement will check if {0,0} is taken or not by using MapSet.member? if it's available, it'll make return {0,0} which will then be used in make_a_move. if it's not available, initial input which always starts at 1, will be incremented by 1 everytime a spot is not available  until a empty spot is found. Hope this clear things up a bit. My goal is to reuse as many functions I have written for the first iteration. Perhaps I should put a comment there to explain how it works?